gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/ProductDependencyLockFile.java

```diff
@@ -16,13 +16,39 @@
 
 package com.palantir.gradle.dist;
 
+import static java.util.stream.Collectors.toList;
+
+import com.google.common.base.Splitter;
 import java.util.List;
 import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 public final class ProductDependencyLockFile {
 
     private static final String HEADER = "# Run ./gradlew --write-locks to regenerate this file\n";
+    private static final String PROJECT_VERSION = "$projectVersion";
+    private static final Pattern LOCK_PATTERN = Pattern.compile(
+            "(?<group>[^:]+):(?<name>[^ ]+) \\((?<min>[^,]+), (?<max>[^\\)]+)\\)");
+
+    public static List<ProductDependency> fromString(String contents, String projectVersion) {
+        return Splitter.on("\n").splitToList(contents).stream()
+                .flatMap(line -> {
+                    Matcher matcher = LOCK_PATTERN.matcher(line);
+                    if (matcher.matches()) {
+                        return Stream.of(new ProductDependency(
+                                matcher.group("group"),
+                                matcher.group("name"),
+                                matcher.group("min").equals(PROJECT_VERSION) ? projectVersion : matcher.group("min"),
```

Comment:
can we have a unit test that covers this `$productVersion` substitution logic pls?